deployment/scripts/upgrade.sh

```diff
@@ -0,0 +1,452 @@
+  if [[ -n "$master" ]]; then
+    waitFor "[[ -n \"\$(oc logs $pod | grep '\[gateway[[:space:]]*\][[:space:]]*\[.*\][[:space:]]*recovered[[:space:]]*\[[[:digit:]]*\][[:space:]]*indices into cluster_state')\" ]]"
+    if [[ $? -ne 0 ]]; then
+      echo "Unable to find log message from gateway that ES recovered its indices within $TIMES seconds"
+      return 1
+    fi
+  else
+    # if we aren't master we should be started by now and should have detected a master
+    if [[ -z "$non_master" ]]; then
+      echo "Node isn't master and was unable to detect master"
+      return 1
+    fi
+  fi
+}
+
+function checkKibanaStarted() {
+
+  local pod=$1
+
+  waitFor "[[ -n \$(oc logs $pod -c kibana | grep 'Listening on 0.0.0.0:5601') ]]"
+  if [[ $? -ne 0 ]]; then
+    echo "Kibana was not able to start up within $TIMES seconds"
+    return 1
+  fi
+}
+
+function scaleDown() {
+
+  # check for a fluentd dc, if there isn't one it was either deleted or using a daemonset
+  local fluentd_dc=(`oc get dc -l logging-infra=fluentd -o jsonpath='{.items[*].metadata.name}'`)
+
+  if [[ -z "$fluentd_dc" ]]; then
+    local selector=$(oc get daemonset -l logging-infra=fluentd -o jsonpath='{.items[*].spec.template.spec.nodeSelector}')
+    if [[ -n "$selector" ]]; then
+      fluentd_nodeselector=$(echo ${selector:4:-1} | sed 's/:/=/g')
+
+      oc delete daemonset logging-fluentd
+    fi
+  else
+    # we are using a deployment config, scale down here
+    for dc in "${fluentd_dc[@]}"; do
+      oc scale --replicas=0 dc $dc
+    done
+  fi
+
+  for pod in $(oc get pods -l component=fluentd -o jsonpath='{.items[*].metadata.name}'); do
+    waitForStop $pod "component=fluentd"
+  done
+
+  local kibana_dc=(`oc get dc -l logging-infra=kibana -o jsonpath='{.items[*].metadata.name}'`)
+  for dc in "${kibana_dc[@]}"; do
+    oc scale --replicas=0 dc $dc
+  done
+  for pod in $(oc get pods -l component=kibana -o jsonpath='{.items[*].metadata.name}'); do
+    waitForStop $pod "component=kibana"
+  done
+
+  local elasticsearch_dc=(`oc get dc -l logging-infra=elasticsearch -o jsonpath='{.items[*].metadata.name}'`)
+  for dc in "${elasticsearch_dc[@]}"; do
+    oc scale --replicas=0 dc $dc
+  done
+  for pod in $(oc get pods -l component=es -o jsonpath='{.items[*].metadata.name}'); do
+    waitForStop $pod "component=es"
+  done
+}
+
+function scaleUp() {
+
+  local elasticsearch_dc=(`oc get dc -l logging-infra=elasticsearch -o jsonpath='{.items[*].metadata.name}'`)
+  for dc in "${elasticsearch_dc[@]}"; do
+    oc scale --replicas=1 dc $dc
+  done
+
+  # scale them all up first and then check that they started up
+  for dc in "${elasticsearch_dc[@]}"; do
+    pod=$(getPodName $dc)
+    waitForStart $pod
+
+    checkESStarted $pod
+  done
+
+  local kibana_dc=(`oc get dc -l logging-infra=kibana -o jsonpath='{.items[*].metadata.name}'`)
+  for dc in "${kibana_dc[@]}"; do
+    oc scale --replicas=1 dc $dc
```

Comment:
maybe put an annotation on the DC when shutting it down :)

On Mon, May 9, 2016 at 11:16 AM, Eric Wolinetz <notifications@github.com>
wrote:

> In deployment/scripts/upgrade.sh
> <https://github.com/openshift/origin-aggregated-logging/pull/119#discussion_r62516848>
> :
>
> > +  local elasticsearch_dc=(`oc get dc -l logging-infra=elasticsearch -o jsonpath='{.items[*].metadata.name}'`)
> > +  for dc in "${elasticsearch_dc[@]}"; do
> > +    oc scale --replicas=1 dc $dc
> > +  done
> > +
> > +  # scale them all up first and then check that they started up
> > +  for dc in "${elasticsearch_dc[@]}"; do
> > +    pod=$(getPodName $dc)
> > +    waitForStart $pod
> > +
> > +    checkESStarted $pod
> > +  done
> > +
> > +  local kibana_dc=(`oc get dc -l logging-infra=kibana -o jsonpath='{.items[*].metadata.name}'`)
> > +  for dc in "${kibana_dc[@]}"; do
> > +    oc scale --replicas=1 dc $dc
>
> I thought about having the deployer do that, I guess since we also have
> this as a start/stop mode it should scale back up as best it can -- keeping
> track of the number of replicas it had before if possible and defaulting to
> replicas=1 otherwise.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin-aggregated-logging/pull/119/files/2d036a3d661ad241f257ea3a09635f8a5da66046#r62516848>
>